### PR TITLE
lmclient: Add --tplsdump option for getting all available templates

### PR DESCRIPTION
**Pull request by vmkononenko**

### CODE REVIEW

File: client/src/lmclient.cpp [hunk 1]
```diff
@@ -416,6 +424,17 @@ void printScales()
...
 
+void dumpTemplates()
+{
+    ChordTplCollection *c = new ChordTplCollection();
+
+    for (uint32_t i = 0; i < c->Size(); i++) {
+        cout << *(c->GetTpl(i)) << endl;
+    }
+
+    free(c);
```
Comment (sa-kib) [inline client/src/lmclient.cpp]: perhaps 
`delete c;
`
or even
`unique_ptr<ChordTplCollection> c;
`
:)

File: client/src/lmclient.cpp [hunk 2]
```diff
@@ -139,7 +144,8 @@ int main(int argc, char* argv[])
         (detectChord && legacy && (winSize || n > 0)) ||
         (printEnvelope && minArgCnt > 3) ||
         (detectBeat && !printTD && minArgCnt > 3) ||
-        (detectBeat && printTD && minArgCnt > 4) || (legacy && minArgCnt == 3))
+        (detectBeat && printTD && minArgCnt > 4) || (legacy && minArgCnt == 3) ||
+        (tplsDump && minArgCnt > 3))
```
Comment (sa-kib) [inline client/src/lmclient.cpp]: `--tplsdump` doesn't require an input file parameter, yet won't work w/o input file:

```
$ ./bin/Debug/lmclient --tplsdump 
Usage:
	lmclient <-t [i] | -f [-p] [-l]> <filename>
[...]
```

```
$ ./bin/Debug/lmclient --tplsdump ../tests/test_sounds/chords_separate/C_chord_construction_guitarlessonworld.wav 
C : 2.3094,0,0,0,1.1547,0,0,1.1547,0,0,0,0,2.3094,0,0,0,2.3094,0,0,2.3094,0,0,0,0
C : 1.26491,0,0,0,1.26491,0,0,1.26491,0,0,0,0,2.52982,0,0,0,2.52982,0,0,2.52982,0,0,0,0
```

File: libmusic/src/chord_tpl.cpp [hunk 3]
```diff
@@ -249,4 +250,14 @@ size_t ChordTpl::SlashSubtypesCnt(chord_quality_t q)
     return chord_qlty_tpls_[q].size();
 }
 
+ostream& operator<<(std::ostream& os, const ChordTpl& tpl)
+{
+    os << Chord(tpl.root_note_, tpl.chord_quality_) << " : ";
+    for (uint8_t i = 0; i < tpl.tpl_.size(); i++) {
```
Comment (sa-kib) [inline libmusic/src/chord_tpl.cpp]:
```suggestion
    for (auto i = 0; i < tpl.tpl_.size(); i++) {
```

File: libmusic/src/chord_tpl.cpp [hunk 4]
```diff
@@ -249,4 +250,14 @@ size_t ChordTpl::SlashSubtypesCnt(chord_quality_t q)
     return chord_qlty_tpls_[q].size();
 }
 
+ostream& operator<<(std::ostream& os, const ChordTpl& tpl)
+{
+    os << Chord(tpl.root_note_, tpl.chord_quality_) << " : ";
```
Comment (sa-kib) [inline libmusic/src/chord_tpl.cpp]: ":" symbol is not CSV-compliant, use "," instead

File: client/src/lmclient.cpp (hunk 1 above)
Comment (vmkononenko) [reply client/src/lmclient.cpp]: what a shame! thanks

File: client/src/lmclient.cpp (hunk 2 above)
Comment (vmkononenko) [reply client/src/lmclient.cpp]: fixed

File: libmusic/src/chord_tpl.cpp (hunk 3 above)
Comment (vmkononenko) [reply libmusic/src/chord_tpl.cpp]: auto goes for a signed int which fails compilation due to comparison with unsigned `tpl.tpl_.size()`

File: client/src/lmclient.cpp [hunk 5]
```diff
@@ -118,6 +120,8 @@ int main(int argc, char* argv[])
             if (i >= argc) { usage(); return 1; }
             refChord = string(argv[i]);
             minArgCnt++;
+        } else if ((strcmp(argv[i], "--tplsdump") == 0)) {
+            tplsDump = true;
```
Comment (sa-kib) [inline client/src/lmclient.cpp]:
```suggestion
            dumpTemplates();
            return 0;
```
Just like we do for  `printScales()`, isn't that simpler ?

File: libmusic/src/chord_tpl.cpp [hunk 6]
```diff
@@ -249,4 +250,14 @@ size_t ChordTpl::SlashSubtypesCnt(chord_quality_t q)
     return chord_qlty_tpls_[q].size();
 }
 
+ostream& operator<<(std::ostream& os, const ChordTpl& tpl)
+{
+    os << Chord(tpl.root_note_, tpl.chord_quality_) << ",";
+    for (uint8_t i = 0; i < tpl.tpl_.size(); i++) {
+        os << tpl.tpl_[i] << ((i < tpl.tpl_.size() - 1) ? "," : "");
+    }
```
Comment (sa-kib) [inline libmusic/src/chord_tpl.cpp]:
```suggestion
    os << Chord(tpl.root_note_, tpl.chord_quality_);
    for (auto &v : tpl.tpl_)
        os << "," << v;

```

isn't that cute?

Comment (vmkononenko) [reply libmusic/src/chord_tpl.cpp]: very cute indeed :)

File: client/src/lmclient.cpp [hunk 7]
```diff
@@ -156,6 +160,7 @@ int main(int argc, char* argv[])
     sf_count_t itemsCnt;
     double *buf;
 
+    inputFilePath = argv[argc - 1];
```
Comment (vmkononenko) [inline client/src/lmclient.cpp]: oops, leftover after review

**Review by sa-kib — APPROVED** [review]
LGTM

**Pull request merged**